rfcs/framework-0057-declarative-python-package-definition.md

```diff
@@ -0,0 +1,174 @@
+
+### Declarative package definition
+
+Below is a commented example how purely declarative ``setup.cfg`` would look like:
+
+```
+# setup.cfg
+[metadata]
+name = invenio-base
+
+# The version number is read by parsing the AST.
+version = attr: invenio_base.__version__
+description = "Base package for building Invenio application factories."
+
+# Files are concatenated as we do today.
+long_description = file: README.rst, CHANGES.rst
+keywords = invenio
+license = MIT
+author = CERN
+author_email = info@inveniosoftware.org
+platforms = any
+url = https://github.com/inveniosoftware/invenio-base
+classifiers =
+    Development Status :: 5 - Production/Stable
+
+[options]
+include_package_data = True
+packages = find:
+
+# Replaces the PyPI classifiers.
+python_requires = >=3.7
+zip_safe = False
+install_requires =
+    blinker>=1.4
+    Flask>=1.0.4,<3.0
+    # importlib needed until Python 3.9 end of life.
+    importlib-metadata>=4.4
+    importlib-resources>=5.0
+    Werkzeug>=1.0.1,<3.0
+
+[options.extras_require]
+tests =
+    pytest-invenio>=1.4.2
+    sphinx>=4.2.0
+
+[options.entry_points]
+console_scripts =
+    inveniomanage = invenio_base.__main__:cli
+flask.commands =
+    instance = invenio_base.cli:instance
+
+[build_sphinx]
+# This option is being deprecated as of Sphinx 5.0 so we
+# might consider removing it already.
+source-dir = docs/
+build-dir = docs/_build
+all_files = 1
+
+[bdist_wheel]
+universal = 1
+
+[pydocstyle]
+add_ignore = D401
+```
+
+#### PyPI classifiers
+
+We suggest removing all classifiers except the classifier related to development status.
+
+- License: The license classifier is already specified in the ``license`` keyword.
+- Python versions: The version classifiers were not kept updated, and where purely metadata, whereas the ``python_requires`` keyword is actionable and easier to keep updated.
+
+#### Version number
+
+The version number has to be moved from ``<package>/version.py`` to ``<package>/__init__.py``. This is because setuptools is only able to extract the version number using the Abstract Syntax Tree (i.e. not executing the code) when located in the ``__init__.py``. If the version number is placed in ``version.py`` setuptools wih import the main package which may lead to imports failing.
+
+#### Extras require
+
+We can no longer programmatically define an ``all`` extra from other extras (like ``tests`` and ``docs``. There's two options:
+
+- Duplicate all extra dependencies in an ``all`` extras (risk of getting outdated).
+- Gather as much as possible into a ``tests`` extras (extra documentation needed).
+
+We suggest as far as possible keeping a single ``all`` extra requires that includes tests and docs requirements. Extra requirements needs to be duplciated into this all requirements, but this means that all packages can mostly be installed via ``pip install -e ``.[all]`` to start development on the package.
```

Comment:
```suggestion
We suggest as far as possible keeping a single ``all`` extra requires that includes tests and docs requirements. Extra requirements needs to be duplicated into this ``all`` requirements, but this means that all packages can mostly be installed via ``pip install -e ``.[all]`` to start development on the package.
```

This is option 1 right? My vote is for that one too: the duplication is for records listed right next to each other so them getting out-of-sync is much less likely.